qf/requests.proto

```diff
@@ -8,8 +8,7 @@ import "qf/types.proto";
 // Message types that are used as requests and responses in API calls.
 
 message CourseSubmissions {
-    Course course                 = 1;  // preloaded assignments
-    repeated EnrollmentLink links = 2;
+    map<uint64, Submissions> submissions = 1;
```

Comment:
I assume the map key is a course ID; We should add a comment `// map: CourseID -> Submissions`

Reply:
The map key is actually EnrollmentID or GroupID based on which SubmissionType is sent in the request. ALL and USER are keyed by EnrollmentID, GROUP is keyed by GroupID.

Reply:
Hmm. I wonder if we should instead make it `EnrollmentID` always. I mean, the `Enrollment` pointed to by an `EnrollmentID` should have a `GroupID` if the enrollment is for a group, or am I just too tired?

Reply:
My thinking was that if we provide `SubmissionType_GROUP`, the returned map *only* contains group submissions. This way I can:
- Range over all groups in a course to easily access the group's submissions: `groups.forEach(group => submissionsByGroup[group.ID])`
- Easily access group submissions for a specific enrollment (lookup by `Enrollment.GroupID`)

This way, if we have three enrollments belonging to the same group, they will all share one entry in our state `submissionsByGroup` map of submissions. 

```ts
// Assignments: 1 2 3 4 5(g) 6(g) 7
// Enrollment IDs: 1, 2, 3, 
// Group IDs: 66 (all enrollments above are in this group)
// Numbers in the array represent which assignment the submission belongs to
Type -> SubmissionType_USER
Submissions -> {1: [1, 2, 3, 4, 7], 2: [1, 2, 3, 4, 7], 3: [1, 2, 3, 4, 7]}

// For this type, the key is GroupID, the two others are keyed by EnrollmentID
Type -> SubmissionType_GROUP
Submissions -> {66: [5, 6]}

// Note submissions for assignments 5 and 6, which are identical for all three enrollments, are duplicated
Type -> SubmissionType_ALL
Submissions -> {1: [1, 2, 3, 4, 5, 6, 7], 2: [1, 2, 3, 4, 5, 6, 7], 3: [1, 2, 3, 4, 5, 6, 7]}
```

Reply:
Okay. I guess we should provide a brief documentation of this above the `CourseSubmissions` message type.